### PR TITLE
fix: ensure ElectronBrowser mojo service is only bound to appropriate render frames

### PR DESCRIPTION
**Pull request by MarshallOfSound**
This ensures that incoming IPC messages from renderer processes that in theory shouldn't have IPC access won't even make it out of their process as the mojo service won't have been bound in the browser process for that render frame.

Notes: no-notes

### CODE REVIEW

File: shell/browser/electron_browser_client.cc [hunk 1]
```diff
@@ -1515,14 +1515,23 @@ void ElectronBrowserClient::
...
-                                           std::move(receiver));
-      },
-      &render_frame_host));
+  auto* contents =
+      content::WebContents::FromRenderFrameHost(&render_frame_host);
+  if (contents) {
+    auto* prefs = WebContentsPreferences::From(contents);
+    if (render_frame_host.IsInPrimaryMainFrame() ||
+        (prefs && prefs->AllowsNodeIntegrationInSubFrames())) {
+      associated_registry.AddInterface(base::BindRepeating(
```
Comment (nornagon) [inline shell/browser/electron_browser_client.cc:1524]: Had a thought since our last conversation about this. There are some non-JS-accessible IPCs that the `ElectronBrowser` interface provides:

```
  OnFirstNonEmptyLayout();

  UpdateDraggableRegions(
    array<DraggableRegion> regions);

  SetTemporaryZoomLevel(double zoom_level);

  [Sync]
  DoGetZoomLevel() => (double result);
```

Do we need any of those on non-main frames?

Comment (MarshallOfSound) [reply shell/browser/electron_browser_client.cc:1524]: @nornagon Maybe we should split those out into a separate mojo interface?

I think they are all legitimately needed in some cases 🤔

Comment (nornagon) [reply shell/browser/electron_browser_client.cc:1524]: yeah i wouldn't be averse to splitting these out. Maybe these ones can stay as `ElectronBrowser` and the rest can go in `ElectronBrowserIpc` or something?

**Review by nornagon — APPROVED** [review]

**Comment by bpasero** [comment]
Could this result in very early `ipcRenderer.invoke` from preload scripts not reaching the browser process? We see VSCode window not getting a result from our call in our preload script when we run with `playwright` Electron smoke tests. This test runner adds `--inspect=0 --remote-debugging-port=0` which maybe is related.

**Comment by bpasero** [comment]
It is hard to say if they work eventually: to clarify: VSCode window refuses to load anything until this early [`ipcRenderer.invoke`](https://github.com/microsoft/vscode/blob/69225f5efec44cd8e26761fca14bd72e9ef6c610/src/vs/base/parts/sandbox/electron-browser/preload.js#L74) is being handled from the browser process (via [`ipcMain.handle`](https://github.com/microsoft/vscode/blob/69225f5efec44cd8e26761fca14bd72e9ef6c610/src/vs/platform/protocol/electron-main/protocolMainService.ts#L142)). 

So in other words, we have no way of knowing if other IPC calls would work since everything stops. I had even experimented with retrying the call 5 seconds later and noticed that even the second attempt was not going through.

**Comment by MarshallOfSound** [comment]
You aren't gonna like it, but if you can get a repro down, that would be helpful.  I'd be surprised if this caused you issues, we tested it extensively.  Can you clarify which renderer is calling `invoke`, is it like a child window, or a web worker / iframe / webview or something weird?

**Comment by bpasero** [comment]
> You aren't gonna like it, but if you can get a repro down, that would be helpful.

Yeah understood, I am trying to create a minimal Electron app that gets run by `playwright` with similar configuration as VSCode (in https://github.com/bpasero/playwright-electron) and still need to see if it reproduces. Unfortunately this is a flaky condition only seen in our CI. We track this issue as https://github.com/microsoft/vscode/issues/146785

> Can you clarify which renderer is calling invoke, is it like a child window, or a web worker / iframe / webview or something weird?

It is a direct `BrowserWindow` created from the browser process. Since we have not yet fully adopted sandbox, it is node.js enabled and not context isolated, but it does use a preload script.
